src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs

```diff
@@ -275,101 +277,151 @@ public static void ValidateHeaderNameCharacters(string headerCharacters)
         {
+            const char lowerCaseChar = (char)0x0020;
+            const ulong lowerCaseULong = 0x0020_0020_0020_0020;
+            const uint lowerCaseUInt = 0x0020_0020;
+            const ushort lowerCaseUShort = 0x0020;
+            // Keep-alive
+            const ulong lowerCaseKeep = 0x0000_0020_0020_0020; // Don't lowercase hyphen
+            const ulong keepChars = 0x002d_0070_0065_0065; // 4 chars "eep-"
+            const ulong alivChars = 0x0076_0069_006c_0061; // 4 chars "aliv"
+            const ushort eChars = 0x0065; // 1 char "e"
+            // Upgrade
+            const ulong pgraChars = 0x0061_0072_0067_0070; // 4 chars "pgra"
+            const uint deChars = 0x0065_0064; // 2 chars "de"
+            // Close
+            const ulong loseChars = 0x0065_0073_006f_006c; // 4 chars "lose"
+
             var connectionOptions = ConnectionOptions.None;
 
             var connectionCount = connection.Count;
             for (var i = 0; i < connectionCount; i++)
             {
-                var value = connection[i];
-                fixed (char* ptr = value)
+                var value = connection[i].AsSpan();
+                while (value.Length > 0)
                 {
-                    var ch = ptr;
-                    var tokenEnd = ch;
-                    var end = ch + value.Length;
-
-                    while (ch < end)
+                    int offset;
+                    char c = '\0';
+                    // Skip any spaces and empty values.
+                    for (offset = 0; offset < value.Length; offset++)
                     {
-                        while (tokenEnd < end && *tokenEnd != ',')
+                        c = value[offset];
+                        if (c == ' ' || c == ',')
                         {
-                            tokenEnd++;
+                            continue;
                         }
 
-                        while (ch < tokenEnd && *ch == ' ')
-                        {
-                            ch++;
-                        }
+                        break;
+                    }
 
-                        var tokenLength = tokenEnd - ch;
+                    // Skip last read char.
+                    offset++;
+                    if ((uint)offset > (uint)value.Length)
+                    {
+                        // Consumed enitre string, move to next.
+                        break;
+                    }
 
-                        if (tokenLength >= 9 && (*ch | 0x20) == 'k')
+                    // Remove leading spaces or empty values.
+                    value = value.Slice(offset);
+                    offset = 0;
+
+                    var potentialConnectionOptions = ConnectionOptions.None;
+                    var byteValue = MemoryMarshal.AsBytes(value);
+                    if ((c | lowerCaseChar) == 'k' && byteValue.Length >= (2 * sizeof(ulong) + sizeof(ushort)))
```

Comment:
Did you try a SIMD-based approach too?
`eep-alive` is too short for a `Vector128<byte>` to compare, but the remainder in the register could be zeroed out before the comparison. Or even check against the full potential option.

This could save some branches and do the comparison in one step instead of breaking them into two.

Reply:
`Vector128<byte>` is not easy to create in a easy performant, safe way from a `Span<byte>`; unless I'm missing something?

Reply:
`MemoryMarshal.Read<Vector128<byte>>(span)`?

Reply:
Not sure its worth it; its quite a chunk
```asm
G_M31882_IG13:
       vmovupd  xmm0, xmmword ptr [rdx]
       mov      rsi, 0x20002000200000
       vmovd    xmm1, rsi
       mov      rsi, 0x20002000200020
       vpinsrq  xmm1, xmm1, xrsi, 1
       vpor     xmm0, xmm0, xmm1
       mov      rsi, 0x2D007000650065
       vmovd    xmm1, rsi
       mov      rsi, 0x760069006C0061
       vpinsrq  xmm1, xmm1, xrsi, 1
       vpcmpeqb xmm0, xmm0, xmm1
       vpmovmskb esi, xmm0
       cmp      esi, 0xFFFF
       sete     sil
       movzx    rsi, sil
       test     esi, esi
       je       G_M31882_IG30
       cmp      r11d, 16
       jb       G_M31882_IG46
```
With this bit being particularly dissapointing
```asm
       cmp      esi, 0xFFFF
       sete     sil
       movzx    rsi, sil
       test     esi, esi
       je       G_M31882_IG30
```

Reply:
> With this bit being particularly dissapointing

I hoped this got fixed with https://github.com/dotnet/runtime/issues/11567 too. Looks like not 😢 

> Not sure its worth it

So it's too much overhead, so I don't believe it's not worth it either. Thanks for trying out!
(If pulling out the "initialization" of the vectors above the loop the overhead will still be there, but payed only once for the price that the code gets less readable...)

Reply:
> If pulling out the "initialization" of the vectors above the loop the overhead will still be there, but payed only once

There's generally only ever one value, so is unlikely to actually loop